docs: added how-tos section with a guide for how to sign a tx (#379)
nervosnetwork/docs.nervos.org: website/docs/how-tos/how-to-sign-a-tx.mdx

---
id: how-to-sign-a-tx
title: How to Sign a Transaction
---

import Tooltip from "@components/Tooltip";

# How to Sign a Transaction

> CKB has two lock scripts with distinct address code hash indexes:
> 
> - `0x00`: Pay to Public Key Hash (P2PKH)
> - `0x01`: Pay to Multisignature (Multisig)

Reviewer: ```suggestion
In CKB, there are 2 primary Lock Scripts with distinct address code hash indexes:

- `0x00`: Pay to Public Key Hash (P2PKH)
- `0x01`: Pay to Multisignature (Multisig)
```

This how-to demonstrates how to sign P2PKH and multisig transactions.

Before you begin, ensure you have a solid grasp of:

- CKB transaction structure, as detailed in [RFC0022: CKB Transaction Structure](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0022-transaction-structure/0022-transaction-structure.md)
- CKB Address Format, as detailed in [RFC0021: CKB Address Format](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0021-ckb-address-format/0021-ckb-address-format.md#short-payload-format)

## Prerequisite Concepts

### Input group

> 

Reviewer: To improve readability, could we put these explanations under the corresponding pseudo code? I feel like this will help users understand the concepts as they follow along with the code.

### WitnessArgs

[`WitnessArgs`](https://github.com/nervosnetwork/ckb/blob/eece5d0c250e0ae0644f9ca9a3f13dab7d7c936c/util/gen-types/schemas/blockchain.mol#L114-L118) allows Lock Script and <Tooltip>Type Script</Tooltip> to read data from different fields of WitnessArgs.

The default P2PKH and the multisig Scripts require user to use `WitnessArgs` on the first `witness` of each group.

:::info

In CKB, `witnesses` is extended to allow users to include any one-time data necessary solely for the transaction’s verification process. However, this flexibility brings a problem, a transaction's Lock Script and Type Script may require different data from the same witness, there is a risk that neither script’s requirement can be met at the same time, potentially causing the transaction permanently locked.

To dismiss this potential conflict, `WitnessArgs` is introduced to allow Lock Script and Type Script to read data from different fields in the `WitnessArgs`.

:::

### Multisig Script

The multisig script hash is a 20-byte blake160 hash, assembled as follows:

```
S | R | M | N | blake160(Pubkey1) | blake160(Pubkey2) | ...
```

> Here, `S`/ `R`/ `M`/ `N` are four single byte unsigned integers, ranging from 0 to 255. `S` is format version, set to 0. `M` and `N` indicate the required number of signatures, M of N, to unlock the cell. `R` specifies that at least R of the provided signatures must match the first R items of the public key list. `blake160 (Pubkey1)` represents the first 160 bits of the blake2b hash of secp256K1 compressed public keys.
> 
> For example, if Alice, Bob, and Cipher collectively control a multisig-locked cell with the unlocking rule "any two of us can unlock the cell, and one must be Cipher’s". The corresponding multisig script is:

Reviewer: ```suggestion
Here, `S`/ `R`/ `M`/ `N` are 4 single-byte unsigned integers, ranging from 0 to 255. 
- `S` is the format version, set to 0. 
- `R` specifies at least R of the provided signatures must match the first R items of the public key list. 
- `M` and `N` indicate the required number of signatures, M of N, to unlock the Cell. 
- `blake160 (Pubkey1)` represents the first 160 bits of the blake2b hash of secp256K1 compressed public keys.

For example, if Alice, Bob, and Cipher collectively control a multisig-locked Cell with the unlocking rule "any two of us can unlock the Cell, and one must be Cipher’s", the corresponding multisig script is:
```

```
0 | 1 | 2 | 3 | Pk_Cipher_h | Pk_Alice_h | Pk_Bob_h
```

To sign a transaction, the multisig script must be revealed in the witness field, followed by the corresponding signatures:

```
multisig_script | Signature1 | Signature2 | ...
```

### Since

The `since` field introduces a timelock constraint, dictating when the multisig lock can be unlocked. It is typically empty, indicating that there is no temporal restriction on unlocking the lock, allowing immediate access.

For an in-depth understanding and detailed exploration, refer to: [RFC0017: Transaction Since Precondition](https://github.com/nervosnetwork/rfcs/blob/5ce42eeadd62c011d5903e82bf3a4304ff1a3f29/rfcs/0017-tx-valid-since/0017-tx-valid-since.md) and [RFC0028: Change Since Relative Timestamp](https://github.com/nervosnetwork/rfcs/blob/5ce42eeadd62c011d5903e82bf3a4304ff1a3f29/rfcs/0028-change-since-relative-timestamp/0028-change-since-relative-timestamp.md?plain=1#L9).

## P2PKH Signing

### Required Arguments

- **Private Key**: a secp256k1 private key
- **Witnesses**: contain transaction signatures

### Pseudo Code

> ```

Reviewer: ```suggestion
```python
```

def sign_tx(pk, tx):

> 		# Group transaction inputs for signing
>     input_groups = compute_input_groups(tx.inputs)

Reviewer: ```suggestion
    # Group transaction inputs for signing
    input_groups = compute_input_groups(tx.inputs)
```

        for witness in tx.witnesses[len(tx.inputs):]
            witness_len_bytes = len(witness).to_le()
            assert(len(witness_len_bytes), 8)
            hasher.update(witness_len_bytes)
            hasher.update(witness)

        # Finalize the hasher to get the signature hash
        sig_hash = hasher.finalize()

        # Sign the transaction with private key
        signature = pk.sign(sig_hash)

        # Replace the dummy lock field with the actual signature in the first witness
        witness_args.lock = signature

        # Serialize the updated witness_args and update the transaction's witnesses list
        tx.witnesses[group_index] = serialize(witness_args)
```

## Multisig Signing

Much like P2PKH transactions, multisig signing is about providing signatures in witnesses to unlock the input. Plus, with each input group isolated, a single transaction can combine both multisig and P2PKH signing methods.

### Required Arguments

- **Private Keys**: a set of secp256k1 private keys
- **Witnesses**: contain transaction signatures
- **Multisig Script**: a Script defining the multisig constraint

### Pseudo Code

> ```

Reviewer: ```suggestion
```python
```

> 
> ```
> def sign_tx(pks, tx, multisig_script):
> 		# Group the transaction inputs for signing
>     input_groups = compute_input_groups(tx.inputs)

Reviewer: ```suggestion

```python
def sign_tx(pks, tx, multisig_script):
    # Group the transaction inputs for signing
    input_groups = compute_input_groups(tx.inputs)
```
